### PR TITLE
feat: add OpenTelemetry metrics reporting

### LINKED ISSUE #93: Export metrics
### Checklist

- [X] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Describe the problem you'd like to have solved

As a consumer of the SDK, I would like to hook it to my dashboards to get data on several metrics.

### Describe the ideal solution

Integrate the OpenTelemetry Python API https://github.com/open-telemetry/opentelemetry-python/tree/main/opentelemetry-api ([API docs](https://opentelemetry-python.readthedocs.io/en/stable/api/index.html))

Expose only from the inner OpenFgaApi. We need to be able to report the following metrics

| Metric Name                     | Type      | Description                                                                     |
|---------------------------------|-----------|---------------------------------------------------------------------------------|
| `fga-client.request.duration`   | Histogram | The total request time for FGA requests                                         |
| `fga-client.query.duration`     | Histogram | The amount of time the FGA server took to process the request                   |
|` fga-client.credentials.request`| Counter   | The total number of times a new token was requested when using ClientCredentials|

And the metrics should have the following attributes associated with them

| Attribute Name                 |  Type     | Description                                                                         |
|--------------------------------|----------|-------------------------------------------------------------------------------------|
| `fga-client.response.model_id` | `string` | The authorization model ID that the FGA server used                                 |
| `fga-client.request.method`    | `string` | The FGA method/action that was performed                                            |
| `fga-client.request.store_id`  | `string` | The store ID that was sent as part of the request                                   |
| `fga-client.request.model_id` | `string` | The authorization model ID that was sent as part of the request, if any                    |
| `fga-client.request.client_id` | `string` | The client ID associated with the request, if any  |
| `fga-client.user`              | `string` | The user that is associated with the action of the request for check and list users |
| `http.status_code `            | `int`    | The status code of the response                                                     |
| `http.method`                  | `string` | The HTTP method for the request                                                     |
| `http.host`                    | `string` | Host identifier of the origin the request was sent to                               |

We should  [create](https://opentelemetry-python.readthedocs.io/en/stable/api/metrics.html#opentelemetry.metrics.get_meter) an overall meter for OpenFgaApi and then create [counters](https://opentelemetry-python.readthedocs.io/en/stable/api/metrics.html#opentelemetry.metrics.Meter.create_counter), [histograms](https://opentelemetry-python.readthedocs.io/en/stable/api/metrics.html#opentelemetry.metrics.Meter.create_histogram) as required for each metric we report.

For the semantically named attributes we can use the semantic conventions package https://pypi.org/project/opentelemetry-semantic-conventions/

### PR DESCRIPTION
**Pull request by evansims**
## Description

Introduces [OpenTelemetry](https://opentelemetry.io/) metrics reporting into the SDK for specific actions. We're intentionally starting with only a few metric events that we can then grow over time.

We don't have any formal documentation for this currently. We'll write this as we roll it out across all SDKs, so here is a summary of the events and the associated attributes:

| Metric Name                     | Type      | Description                                                                     |
|---------------------------------|-----------|---------------------------------------------------------------------------------|
| `fga-client.request.duration`   | Histogram | The total request time for FGA requests                                         |
| `fga-client.query.duration`     | Histogram | The amount of time the FGA server took to process the request                   |
|` fga-client.credentials.request`| Counter   | The total number of times a new token was requested when using ClientCredentials|

| Attribute Name                 |  Type     | Description                                                                         |
|--------------------------------|----------|-------------------------------------------------------------------------------------|
| `fga-client.response.model_id` | `string` | The authorization model ID that the FGA server used                                 |
| `fga-client.request.method`    | `string` | The FGA method/action that was performed                                            |
| `fga-client.request.store_id`  | `string` | The store ID that was sent as part of the request                                   |
| `fga-client.request.model_id` | `string` | The authorization model ID that was sent as part of the request, if any                    |
| `fga-client.request.client_id` | `string` | The client ID associated with the request, if any  |
| `fga-client.user`              | `string` | The user that is associated with the action of the request for check and list users |
| `http.status_code `            | `int`    | The status code of the response                                                     |
| `http.method`                  | `string` | The HTTP method for the request                                                     |
| `http.host`                    | `string` | Host identifier of the origin the request was sent to                               |

We're implementing this using `opentelemetry-api,` which means that all actions will be a no-op unless an `opentelemetry-sdk` (or equivalent) instance is configured within an application. This is currently in draft as we work through the implementation across SDKs, but we're looking for any feedback folks may have!

An example implementation is provided and can be found under `opentelemetry` in the `example` directory of this repository.

## References

Closes #93 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by evansims** [comment]
One important thing to note, I've opted not to use the `opentelemetry-semantic-conventions` package here for a number of reasons.

1. I noticed its definitions were incomplete and missing a number of basic attribute names we were expecting to be present.
2. Several names had differing values from other languages' versions of the package.

It didn't make sense to implement the package in such an unfinished state here.

**Review by ewanharris — APPROVED** [review]
Looks good to me!

I really like the abstraction you built around the telemetry, definitely will look to do that in JS as we roll out the metrics/span work across more of the API

File: example/opentelemetry/main.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import asyncio
+import os
+import sys
+from operator import attrgetter
+from random import randint
+from typing import Any
+
+from dotenv import load_dotenv
+from opentelemetry import metrics
+from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
```
Comment (ewanharris) [inline example/opentelemetry/main.py:10]: Ah sorry I overlooked this running the example, here we specify `opentelemetry.exporter.otlp.proto.grpc.metric_exporter` but the requirements specify `opentelemetry-exporter-otlp-proto-http`

Should we switch this to the http one or add the grpc one to the requirements?

Comment (evansims) [reply example/opentelemetry/main.py:10]: Ah whoop, good catch, updated! I swapped between them a few times testing and guessed I missed updating that 😆

**Review by ewanharris — APPROVED** [review]

**Pull request merged**